File: btree.js
/**
 * @compiler Bridge.NET 17.4.0
 */
Bridge.assembly("Demo", function ($asm, globals) {
    "use strict";

    Bridge.define("Btree.Bnode", {
        fields: {
            n: 0,
            size: 0,
            keys: null,
            children: null,
            parent: null
        },
        ctors: {
            ctor: function (m) {
                this.$initialize();
                this.size = (m - 1) | 0;
                this.n = 0;
                this.keys = System.Array.init(((this.size + 1) | 0), 0, System.Int32);
                this.children = System.Array.init(((this.size + 2) | 0), null, Btree.Bnode);
                this.children[System.Array.index(0, this.children)] = null;
                this.parent = null;
            }
        },
        methods: {
            insert: function (key) {
                if (this.n === 0) {
                    return this.insertFirstKey(key);
                }
                if (key > this.keys[System.Array.index(((this.n - 1) | 0), this.keys)]) {
                    return this.insertLast(key);
                }
                return this.insertOrder(key);
            },
            insertOrder: function (key) {
                var position = this.findPosition(key);
                for (var i = this.n; i > position; i = (i - 1) | 0) {
                    this.children[System.Array.index(((i + 1) | 0), this.children)] = this.children[System.Array.index(i, this.children)];
                    this.keys[System.Array.index(i, this.keys)] = this.keys[System.Array.index(((i - 1) | 0), this.keys)];
                }
                this.keys[System.Array.index(position, this.keys)] = key;
                this.children[System.Array.index(((position + 1) | 0), this.children)] = null;
                this.n = (this.n + 1) | 0;
                return position;
            },
            insertFirstKey: function (key) {
                this.keys[System.Array.index(0, this.keys)] = key;
                this.children[System.Array.index(1, this.children)] = null;
                this.n = 1;
                return 0;
            },
            insertLast: function (key) {
                this.keys[System.Array.index(Bridge.identity(this.n, (this.n = (this.n + 1) | 0)), this.keys)] = key;
                this.children[System.Array.index(this.n, this.children)] = null;
                return ((this.n - 1) | 0);
            },
            findPosition: function (key) {
                var position = System.Array.binarySearch(this.keys, 0, this.n, key);
                if (position < 0) {
                    position = ~position;
                }
                return position;
            },
            delete: function (position) {
                for (var i = (position + 1) | 0; i < this.n; i = (i + 1) | 0) {
                    this.keys[System.Array.index(((i - 1) | 0), this.keys)] = this.keys[System.Array.index(i, this.keys)];
                    this.children[System.Array.index(i, this.children)] = this.children[System.Array.index(((i + 1) | 0), this.children)];
                }
                this.n = (this.n - 1) | 0;
                return this.n;
            },
            findChild: function (child) {
                for (var i = 0; i <= this.n; i = (i + 1) | 0) {
                    if (Bridge.referenceEquals(this.children[System.Array.index(i, this.children)], child)) {
                        return i;
                    }
                }
                return -1;
            }
        }
    });

    Bridge.define("Btree.Btree", {
        fields: {
            bt: null,
            m: 0
        },
        ctors: {
            ctor: function (m) {
                this.$initialize();
                this.bt = new Btree.Bnode(3);
                this.m = m;
            }
        },
        methods: {
            insert: function (root, key) {
                if (root.v == null) {
                    root.v = new Btree.Bnode(this.m);
                    root.v.insert(key);
                } else {
                    var current = root.v;
                    while (true) {
                        if (current.children[System.Array.index(0, current.children)] == null) {
                            var position = current.insert(key);
                            while (current.n > current.size) {
                                var result = this.split(root, current);
                                var newNode = Bridge.cast(result[System.Array.index(0, result)], Btree.Bnode);
                                key = Bridge.cast(result[System.Array.index(1, result)], System.Int32);
                                current = current.parent;
                                position = current.insert(key);
                                current.children[System.Array.index(((position + 1) | 0), current.children)] = newNode;
                            }
                            break;
                        }
                        current = current.children[System.Array.index(current.findPosition(key), current.children)];
                    }
                }
            },
            delete: function (root, key) {
                var $t, $t1, $t2;
                var current = this.find(root.v, key);
                if (current != null) {
                    if (($t = current.multiNode.children)[System.Array.index(current.position, $t)] != null) {
                        var predecessor = ($t1 = current.multiNode.children)[System.Array.index(current.position, $t1)];
                        while (predecessor.children[System.Array.index(predecessor.n, predecessor.children)] != null) {
                            predecessor = predecessor.children[System.Array.index(predecessor.n, predecessor.children)];
                        }
                        ($t2 = current.multiNode.keys)[System.Array.index(current.position, $t2)] = predecessor.keys[System.Array.index(((predecessor.n - 1) | 0), predecessor.keys)];
                        current = new Btree.Node(predecessor, ((predecessor.n - 1) | 0));
                    }
                    current.multiNode.delete(current.position);
                    var minKeys = (Bridge.Int.div(current.multiNode.size, 2)) | 0;
                    if (Bridge.referenceEquals(current.multiNode, root.v) && current.multiNode.n === 0) {
                        root.v = null;
                    }
                    while (root.v != null && !Bridge.referenceEquals(current.multiNode, root.v) && current.multiNode.n < minKeys) {
                        this.rebalanceAfterDeletion(root, Bridge.ref(current, "multiNode"), minKeys);
                    }
                }
            },
            inorder: function (root) {
                if (root.v != null) {
                    for (var i = 0; i < root.v.n; i = (i + 1) | 0) {
                        this.inorder(Bridge.ref(root.v.children, i));
                        System.Console.WriteLine("Keys : " + root.v.keys[System.Array.index(i, root.v.keys)] + " n : " + root.v.n + " parent value : " + (root.v.parent==null ? "null" : root.v.parent.keys[0]));
                    }
                    this.inorder(Bridge.ref(root.v.children, root.v.n));
                }
            },
            split: function (root, node) {
                var mid = (Bridge.Int.div(node.n, 2)) | 0;
                if (node.parent == null) {
                    root.v = new Btree.Bnode(this.m);
                    root.v.children[System.Array.index(0, root.v.children)] = node;
                    node.parent = root.v;
                }
                var newNode = new Btree.Bnode(this.m);
                newNode.parent = node.parent;
                var nk = 0;
                for (var k = (mid + 1) | 0; k < node.n; k = (k + 1) | 0, nk = (nk + 1) | 0) {
                    newNode.insert(node.keys[System.Array.index(k, node.keys)]);
                    this.moveParent(newNode, nk, node.children[System.Array.index(k, node.children)]);
                }
                this.moveParent(newNode, nk, node.children[System.Array.index(node.n, node.children)]);
                node.n = mid;
                return System.Array.init([newNode, node.keys[System.Array.index(mid, node.keys)]], System.Object);
            },
            moveParent: function (parent, position, child) {
                parent.children[System.Array.index(position, parent.children)] = child;
                if (child != null) {
                    child.parent = parent;
                }
            },
            find: function (root, key) {
                var current = root;
                while (current != null) {
                    var position = current.findPosition(key);
                    if (position < current.n && current.keys[System.Array.index(position, current.keys)] === key) {
                        return new Btree.Node(current, position);
                    } else {
                        current = current.children[System.Array.index(position, current.children)];
                    }
                }
                return null;
            },
            rebalanceAfterDeletion: function (root, current, minKeys) {
                var parent = current.v.parent;
                var i = parent.findChild(current.v);
                if (i > 0 && parent.children[System.Array.index(((i - 1) | 0), parent.children)].n > minKeys) {
                    this.rightRotate(parent, ((i - 1) | 0));
                    return true;
                }
                if (i > parent.n && parent.children[System.Array.index(((i + 1) | 0), parent.children)].n > minKeys) {
                    this.leftRotate(parent, i);
                    return true;
                }
                if (i > 0) {
                    current.v = this.merge(parent, ((i - 1) | 0));
                } else {
                    current.v = this.merge(parent, i);
                }
                if (parent.n === 0 && Bridge.referenceEquals(parent, root.v)) {
                    root.v = current.v;
                    current.v.parent = null;
                } else {
                    current.v = parent;
                }
                return false;
            },
            leftRotate: function (parent, position) {
                var left = parent.children[System.Array.index(position, parent.children)];
                var right = parent.children[System.Array.index(((position + 1) | 0), parent.children)];
                var newPosition = left.insert(parent.keys[System.Array.index(position, parent.keys)]);
                this.moveParent(left, ((newPosition + 1) | 0), right.children[System.Array.index(0, right.children)]);
                parent.keys[System.Array.index(position, parent.keys)] = right.keys[System.Array.index(0, right.keys)];
                right.children[System.Array.index(0, right.children)] = right.children[System.Array.index(1, right.children)];
                right.delete(0);
            },
            rightRotate: function (parent, position) {
                var left = parent.children[System.Array.index(position, parent.children)];
                var right = parent.children[System.Array.index(((position + 1) | 0), parent.children)];
                var newPosition = right.insert(parent.keys[System.Array.index(position, parent.keys)]);
                right.children[System.Array.index(1, right.children)] = right.children[System.Array.index(0, right.children)];
                this.moveParent(right, newPosition, left.children[System.Array.index(left.n, left.children)]);
                parent.keys[System.Array.index(position, parent.keys)] = left.keys[System.Array.index(((left.n - 1) | 0), left.keys)];
                left.n = (left.n - 1) | 0;
            },
            merge: function (parent, position) {
                var left = parent.children[System.Array.index(position, parent.children)];
                var right = parent.children[System.Array.index(((position + 1) | 0), parent.children)];
                var newPosition = left.insert(parent.keys[System.Array.index(position, parent.keys)]);
                this.moveParent(left, ((newPosition + 1) | 0), right.children[System.Array.index(0, right.children)]);
                for (var i = 0; i < right.n; i = (i + 1) | 0) {
                    newPosition = left.insert(right.keys[System.Array.index(i, right.keys)]);
                    this.moveParent(left, ((newPosition + 1) | 0), right.children[System.Array.index(((i + 1) | 0), right.children)]);
                }
                parent.delete(position);
                return left;
            }
        }
    });

    Bridge.define("Btree.Node", {
        fields: {
            multiNode: null,
            position: 0
        },
        ctors: {
            ctor: function (multiNode, position) {
                this.$initialize();
                this.multiNode = multiNode;
                this.position = position;
            }
        }
    });

    Bridge.define("Btree.Program", {
        main: function Main(args) {
            var bt = new Btree.Btree(3);
            bt.insert(Bridge.ref(bt, "bt"), 20);
            bt.insert(Bridge.ref(bt, "bt"), 10);
            bt.insert(Bridge.ref(bt, "bt"), 5);
            /*
            bt.insert(Bridge.ref(bt, "bt"), 7);
            bt.insert(Bridge.ref(bt, "bt"), 9);
            bt.insert(Bridge.ref(bt, "bt"), 6);
            */
            System.Console.WriteLine("Print!");
            bt.inorder(Bridge.ref(bt, "bt"));
            /*
            bt.delete(Bridge.ref(bt, "bt"), 7);
            System.Console.WriteLine("After delete 7!");
            bt.inorder(Bridge.ref(bt, "bt"));
            */
        }
    });
});

File: Btree/Btree/Btree.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Btree {
    class Btree {
        public Bnode root;
        public int m;
        // CODE
        #region
        public Btree(int m) {
            root = new Bnode(m);//,0);
            this.m = m;
        }

        public void insert(ref Bnode root, int key) {
            int currentDepth = 0;

            if (root == null) {
                root = new Bnode(m);//,0);
                root.insert(key);
            } else {
                Bnode current = root;
                while (true) {
                    if (current.children[0] == null) { //insert must be in leaf
                        int position = current.insert(key);
                        while (current.n > current.size) { //overflow
                                                           //split the node
                            Object[] result = split(ref root, current, currentDepth);
                            //try to insert middle key to parent
                            Bnode newNode = (Bnode)result[0];
                            key = (int)result[1]; //separator
                            current = current.parent;
                            position = current.insert(key);
                            current.children[position + 1] = newNode;
                        }
                        break;
                    }
                    current = current.children[current.findPosition(key)];

                    currentDepth++;
                }
            }
        }

        public void inorder(ref Bnode root) {
            if (root != null) {
                for (int i = 0; i < root.n; i++) {
                    inorder(ref root.children[i]);
                    Console.WriteLine("Keys : " + root.keys[i] + " n : " + root.n + " parent.keys[0] : " + (root.parent == null ? "-" : root.parent.keys[0].ToString())); // used when you're not sure visualization is correct
                }
                inorder(ref root.children[root.n]);
            }
        }

        public void delete(ref Bnode root, int key) {
            Node current = find(root, key);
            if (current != null) {
                if (current.multiNode.children[current.position] != null) {
                    //current is not leaf
                    //..
                    //find in-order predecessor
                    Bnode predecessor =
                    current.multiNode.children[current.position];
                    while (predecessor.children[predecessor.n] != null) {
                        predecessor = predecessor.children[predecessor.n];
                    }
                    //...
                    //replace with the predecessor
                    current.multiNode.keys[current.position] =
                    predecessor.keys[predecessor.n - 1];
                    //try to delete leaf
                    current = new Node(predecessor, predecessor.n - 1);
                }
                //delete leaf
                //...
                current.multiNode.delete(current.position);
                int minKeys = current.multiNode.size / 2;
                if (current.multiNode == root && current.multiNode.n == 0)
                    root = null;
                while (root != null && current.multiNode != root && current.multiNode.n < minKeys) //underflow
                    rebalanceAfterDeletion(ref root, ref current.multiNode, minKeys);
            }
        }

        // other
        Object[] split(ref Bnode root, Bnode node, int currentDepth) {
            int mid = node.n / 2;
            if (node.parent == null) {
                root = new Bnode(m);//, currentDepth);
                root.children[0] = node;
                node.parent = root;
            }
            Bnode newNode = new Bnode(m);//, currentDepth);
            newNode.parent = node.parent;
            int nk = 0;
            for (int k = mid + 1; k < node.n; k++, nk++) {
                newNode.insert(node.keys[k]);
                moveParent(newNode, nk, node.children[k]);
            }
            moveParent(newNode, nk, node.children[node.n]);
            node.n = mid;
            return new Object[] { newNode, node.keys[mid] };
        }

        void moveParent(Bnode parent, int position, Bnode child) {
            parent.children[position] = child;
            if (child != null)
                child.parent = parent;
        }

        public Node find(Bnode root, int key) {
            Bnode current = root;
            while (current != null) {
                int position = current.findPosition(key);
                if (position < current.n && current.keys[position] == key)
                    return new Node(current, position); //found
                else
                    current = current.children[position];
            }
            return null; //not found
        }

        bool rebalanceAfterDeletion(ref Bnode root, ref Bnode current, int minKeys) {
            Bnode parent = current.parent;
            int i = parent.findChild(current);
            //case 1: look at left sibling
            if (i > 0 && parent.children[i - 1].n > minKeys) {
                rightRotate(parent, i - 1);
                return true;
            }
            //case 2: look at right sibling
            if (i > parent.n && parent.children[i + 1].n > minKeys) {
                leftRotate(parent, i);
                return true;
            }
            //...
            //merge
            if (i > 0)
                current = merge(parent, i - 1); //case 3: merge to left sibling
            else
                current = merge(parent, i); //case 3: merge to right sibling
            if (parent.n == 0 && parent == root) {
                root = current;
                current.parent = null;
            } else
                current = parent;
            return false;
        }
        
        void leftRotate(Bnode parent, int position) {
            Bnode left = parent.children[position];
            Bnode right = parent.children[position + 1];
            int newPosition = left.insert(parent.keys[position]);
            moveParent(left, newPosition + 1, right.children[0]);
            parent.keys[position] = right.keys[0];
            right.children[0] = right.children[1];
            right.delete(0);
        }

        void rightRotate(Bnode parent, int position) {
            Bnode left = parent.children[position];
            Bnode right = parent.children[position + 1];
            int newPosition = right.insert(parent.keys[position]);
            right.children[1] = right.children[0];
            moveParent(right, newPosition, left.children[left.n]);
            parent.keys[position] = left.keys[left.n - 1];
            left.n--;
        }

        Bnode merge(Bnode parent, int position) {
            Bnode left = parent.children[position];
            Bnode right = parent.children[position + 1];
            int newPosition = left.insert(parent.keys[position]);
            moveParent(left, newPosition + 1, right.children[0]);
            for (int i = 0; i < right.n; i++) {
                newPosition = left.insert(right.keys[i]);
                moveParent(left, newPosition + 1, right.children[i + 1]);
            }
            parent.delete(position);
            return left;
        }
        #endregion

        // VISUALIZATION
        #region
        List<FakeBNode> fakeBNodes;
        int[][] traverseIndex;
        public int maxDepth;

        int nodeCount;
        int nodeCtr;
        public List<FakeBNode> getFakeBNodes() {
            fakeBNodes = new List<FakeBNode>();
            maxDepth = 0;
            nodeCount = 0;
            nodeCtr = 0;

            getMaxDepth(ref root);
            getNodeCount(ref root);
            traverseIndex = new int[nodeCount][];

            getFakeBNodes(ref root);
            fakeBNodes.Sort(new CompareFakeBNode());
            return fakeBNodes;
        }

        public void getMaxDepth(ref Bnode root, int depth = 0) {
            if (root != null) {
                if (depth > maxDepth) {
                    maxDepth = depth;
                }

                for (int i = 0; i < root.n; i++) {
                    getMaxDepth(ref root.children[i], depth + 1);
                }
            }
        }

        public void getNodeCount(ref Bnode root) {
            if (root != null) {
                nodeCount++;
                for (int i = 0; i < root.n + 1; i++) {
                    getNodeCount(ref root.children[i]);
                }
            }
        }
        
        public void getFakeBNodes(ref Bnode root, int parentIndex = -1, int traverseWay = -1, int depth = 0) {
            if (root != null) {
                int nodeIndex = nodeCtr;
                nodeCtr++;

                // set traverse index
                if (parentIndex == -1) {
                    traverseIndex[nodeIndex] = new int[1];
                    traverseIndex[nodeIndex][0] = -1;
                } else {
                    if (traverseIndex[parentIndex][0] == -1) {
                        traverseIndex[nodeIndex] = new int[1];
                        traverseIndex[nodeIndex][0] = traverseWay;
                    } else {
                        traverseIndex[nodeIndex] = new int[traverseIndex[parentIndex].Count() + 1];
                        for (int i = 0; i < traverseIndex[parentIndex].Count(); i++) {
                            traverseIndex[nodeIndex][i] = traverseIndex[parentIndex][i];
                        }
                        traverseIndex[nodeIndex][traverseIndex[nodeIndex].Count() - 1] = traverseWay;
                    }
                }

                // get childCount & recursive
                int childCount = 0;
                for (int i = 0; i < root.n + 1; i++) {
                    if (root.children[i] != null) {
                        childCount++;
                        getFakeBNodes(ref root.children[i], nodeIndex, i, depth + 1);
                    }
                }

                // add to fakeBNodes
                fakeBNodes.Add(new FakeBNode(depth, root.n, root.keys, traverseIndex[nodeIndex], childCount));
            }
        }

        public List<int> getFindTraverseIndex(Bnode root, int key) {
            Bnode current = root;
            List<int> searchTraverseIndex = new List<int>();
            while (current != null) {
                int position = current.findPosition(key);
                if (position < current.n && current.keys[position] == key) {
                    if (current == this.root) {
                        searchTraverseIndex.Add(-1);
                    }
                    return searchTraverseIndex; //found
                } else {
                    current = current.children[position];
                    searchTraverseIndex.Add(position);
                }
            }
            return null; //not found
        }
        #endregion
    }
}
